File: js/src/0024-swapPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
// @lc app=leetcode.cn id=24 lang=javascript
// 60 ms, 在所有 JavaScript 提交中击败了70.94%
const swapPairs = function(head) {
  const changeNode = (node) => {
    if (!node) {
      return node;
    }
    if (node && !node.next) {
      return node;
    }
    // 有 node 和 node.next
    const a = node;
    const b = node.next;
    const c = node.next.next;
    // 前后交换节点，改变指针
    a.next = changeNode(c);
    b.next = a;
    return b;
  };
  return changeNode(head);
};

export { swapPairs };

File: others/0024-swapPairs.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
// [24] 两两交换链表中的节点
// Your runtime beats 100 % of c submissions
struct ListNode* swapPairs(struct ListNode* head){
  // 没有节点或者没有下一个节点
  if (head == NULL) {
    return head;
  }
  if (head && head->next == NULL) {
    return head;
  }
  // head 是当前节点，a 和 b 分别下一个和下下一个
  struct ListNode* a = head->next;
  struct ListNode* b = head->next->next;
  // 改变指针，调换 head 和 节点 a，并递归后面的节点
  head->next = swapPairs(b);
  a->next = head;
  return a;
}
